### ISSUE TITLE
Parallel training ends with an error 

### ISSUE THREAD

**Issue by gonanc13** [jerabaul29/Cylinder2DFlowControlDRLParallel#2]
Hi,

when I launch a parallel learning session on a container spun up from the Docker image provided in the Readme, all the specified servers are set up and learning takes place as expected. However as the last episode of the last training batch is simulated, the execution terminates with the following error:

```
    RemoteEnvironmentServer(tensorforce_environment, host=host, port=port)
  File "/home/fenics/local/Cylinder2DFlowControlDRLParallel/Cylinder2DFlowControlWithRL/env_8/../RemoteEnvironmentServer.py", line 52, in __init__
    request, data = EchoServer.decode_message(msg, verbose=self.verbose)
    response = self.handle_message(message)  # this is given by the EchoServer base class
  File "/home/fenics/local/Cylinder2DFlowControlDRLParallel/Cylinder2DFlowControlWithRL/env_8/../echo_server.py", line 72, in handle_message
EOFError: Ran out of input
    request, data = EchoServer.decode_message(msg, verbose=self.verbose)
  File "/home/fenics/local/Cylinder2DFlowControlDRLParallel/Cylinder2DFlowControlWithRL/env_8/../echo_server.py", line 35, in decode_message
    msg = pickle.loads(msg)
EOFError: Ran out of input
```

I was wondering if this is an error that you had encountered in the past or maybe I am doing something wrong?

Thank you very much for your work.

**Comment by jerabaul29** [comment]
Hi,

Sorry for the late reply, I had missed your issue. Was not watching the repo, doing it now, should not miss again ^^ .

This code and docker have been used by many people across several universities, so I am quite confident it should work. It think the problem is probably how you start the training. What command do you use? You should issue something like:

```
bash script_launch_parallel.sh session_name first_port num_servers
```

where ```session_name``` is any string you want (like, ```test_session```), first_port is the first port to use (make sure this port and all following are available, if necessary, play a bit around with the number until you find some; like ```6000``` uses to be ok), and ```num_servers``` how many servers, ie cfd simulations, you want to use.

Can you:

- try that
- let me know if it works
- if it does not work, show me a screenshot

**Comment by gonanc13** [comment]
Hi, 

this is the command I'm using. Specifically I am issuing 

`bash script_launch_parallel.sh training 8000 20`

to launch 20 parallel environments. 

The servers are launched and training seems to run normally. However, when the progress bar reaches 100% and the number of episodes set by argument num_episodes of runner.run() have been simulated, it seems like new simulations start in each server. Shortly after, the error I sent in my previous message appears.

If, for example, num_episodes was 400, when I then look at env_#/saved_models/output.csv I find the averages of 22 episodes per environment instead of 20 as I think it should be for 20 parallel environments.

**Comment by jerabaul29** [comment]
Aaah, ok, I see. Yes, this is not really a problem: basically, I never cared to implement 'gracious termination' of the training from the simulations side. So basically it should be that:

- the training happens as it should, without problems
- at the end of the training, the ANN stops communicating with the simulations
- at this point, the simulations 'crash' and complain, but the ANN continues to do his work, maybe dump the network, etc.

So I agree that this is a bit ugly / the code does not clean after itself, but it should have no consequence other than those ugly messages, right? You can just kill stuff by hand or with a script when you are done.
Indeed, this means that a last episode may be started, so 21, and then there may be some off-by-one error here or there, which should have no importance in practice so was never thoroughly debugged. Also, sometimes some simulations progress faster than others, in which case, maybe you actually have 22 episodes one one machine, but only 19 on another - which is quite fine :) .

Can you confirm that:

- I understand your problem right
- you agree
- that does not impact your training etc?

To summarize, things are not perfect, especially small details that do not have an impact on the training are not too polished. Should not have anything to day though. Do you agree? :)

Great you were able to run your trainings etc :) .

**Comment by jerabaul29** [comment]
PS: if you want to do very large trainings you can also train across machines. See the explanations I have been giving there:

https://github.com/jerabaul29/Cylinder2DFlowControlDRLParallel/issues/3

**Comment by gonanc13** [comment]
I see! 

This is what I thought, I just wanted to make sure there wasn't a deeper problem in the way I was doing things that was causing this error. 

Indeed, this does not impact the training as the episode returns increase with the number of episodes and, once trained,  the agent manages to reduce the drag when a single episode is ran. 

And I agree, this 'ugly' termination is not a problem at all if the code still serves its purpose!

I will take a look at #3 to see if I can make use of it to speed up my trainings.

Thanks very much for your response and great work!

**Comment by jerabaul29** [comment]
Yes, no worries at all, always good to ask :) Thanks! Feel free to discuss on #3 if anything around the training through ssh is unclear :) .
